### PR TITLE
Ensure that data objects cannot get stuck in locked status (4-2-stable)

### PR DESCRIPTION
**Pull request by alanking**
Addresses https://github.com/irods/irods/issues/6154
Addresses https://github.com/irods/irods/issues/6378
Addresses https://github.com/irods/irods/issues/6479
Addresses https://github.com/irods/irods/issues/6880
Addresses https://github.com/irods/irods/issues/6882
Addresses https://github.com/irods/irods/issues/6886

Companion PR: https://github.com/irods/irods_resource_plugin_s3/pull/2097

Cherry-picked from https://github.com/irods/irods/pull/6884

Core tests are running now. Need to make sure S3 plugin tests pass before merging

There are some differences between this branch and main/4-3-stable. Specifically, the agent cleanup logic is in a completely different file, so the cherry-pick was rather manual in that part. Please review the first (that is, earliest/base) commit carefully.

### CODE REVIEW

**Comment by alanking** [comment]
Force-push was just a rebase. Unit and core tests are still passing. Still trying to figure out what's going on with S3

File: server/api/src/rsDataObjPut.cpp [hunk 1]
```diff
@@ -328,10 +328,37 @@ namespace
 
             // close the replica, free L1 descriptor
             if (const int ec = irods::close_replica_without_catalog_update(_comm, fd, preserve_rst); ec < 0) {
+                // Closing the replica failed. We need to clean up everything and unlock the data object so that this
+                // data object is left in a workable state.
                 irods::log(LOG_ERROR, fmt::format(
                     "[{}:{}] - error closing replica; ec:[{}]",
                     __FUNCTION__, __LINE__, ec));
 
+
```
Comment (korydraughn) [inline server/api/src/rsDataObjPut.cpp:337]: Found an extra line

Comment (alanking) [reply server/api/src/rsDataObjPut.cpp]: Done

**Review by korydraughn — COMMENTED** [review]
Code changes look fine.

Are we still investigating S3 for this PR?

**Comment by alanking** [comment]
Nope, all has been resolved in irods/irods_resource_plugin_s3#2097.

**Review by korydraughn — APPROVED** [review]
Add pounds and merge when ready.

**Comment by korydraughn** [comment]
Don't forget to double check the issue numbers if you haven't already.

**Pull request merged**